telex/telex.go

```diff
@@ -0,0 +1,219 @@
+// package telex executes process and streams process output to clients
+package telex
+
+import (
+	"context"
+	"errors"
+	"os/exec"
+	"sync"
+
+	"github.com/google/shlex"
+	"github.com/google/uuid"
+
+	"github.com/qrdl/telex/spooler"
+)
+
+// Process status codes
+const (
+	StatusInvalid = iota
+	StatusRunning
+	StatusCompleted
+	StatusFailed
+	StatusKilled
+)
+
+// Output stream types
+const (
+	_ = iota
+	StreamStdout
+	StreamStderr
+)
+
+const invalidHandle = ""
+
+// cache of all executed and currently executing processes
+type process struct {
+	sync.Mutex
+	status      int
+	owner       string
+	exitcode    int
+	stdout      *spooler.Spooler
+	stderr      *spooler.Spooler
+	killChannel chan bool
+}
+
+type jobLog struct {
+	sync.RWMutex // mutex to allow safely adding new records to runLog
+	log   map[uuid.UUID]*process
+}
+
+var jobs = jobLog{log: map[uuid.UUID]*process{}}
+
+// StartProcess starts execution of provided command, returns process handle or error.
+func StartProcess(cmdLine string, user string) (string, error) {
+	params, err := shlex.Split(cmdLine)
+	if err != nil {
+		return invalidHandle, err
+	}
+
+	var proc *exec.Cmd
+	if len(params) > 1 {
+		proc = exec.Command(params[0], params[1:]...)
+	} else {
+		proc = exec.Command(params[0])
+	}
+
+	stdout, err := proc.StdoutPipe()
+	if err != nil {
+		return invalidHandle, err
+	}
+	stderr, err := proc.StderrPipe()
+	if err != nil {
+		return invalidHandle, err
+	}
+
+	err = proc.Start()
+	if err != nil {
+		return invalidHandle, err
+	}
+
+	// process started - add entry to run log
+	id := uuid.New()
+	stdoutSpooler, err := spooler.NewSpooler(stdout)
+	if err != nil {
+		return invalidHandle, err
+	}
+	stderrSpooler, err := spooler.NewSpooler(stderr)
+	if err != nil {
+		return invalidHandle, err
+	}
+	job := &process{
+		status:      StatusRunning,
+		owner:       user,
+		stdout:      stdoutSpooler,
+		stderr:      stderrSpooler,
+		killChannel: make(chan bool)}
+
+	jobs.Lock()
+	defer jobs.Unlock()
+	jobs.log[id] = job
+
+	go execute(proc, job)
+	return id.String(), nil
+}
+
+// GetOutputStream returns channel the client can read process output from.
+// The channel will be closed when all output of the process is delivered.
+// Client can cancel the stream using context.
+// It is a responsibility of the caller to check whether process belongs to the user.
+func GetOutputStream(ctx context.Context, handle string, streamType int) (<-chan string, error) {
+	if streamType < StreamStdout || streamType > StreamStderr {
+		return nil, errors.New("invalid stream type")
+	}
+	id, err := uuid.Parse(handle)
+	if err != nil {
+		return nil, err
+	}
+
+	jobs.RLock()
+	job, ok := jobs.log[id]
+	jobs.RUnlock()
+	if !ok {
+		return nil, errors.New("invalid handle")
+	}
+
+	job.Lock()
+	defer job.Unlock()
+	if streamType == StreamStdout {
+		return job.stdout.Attach(ctx), nil
+	}
```

Comment:
Since you need a write lock anyway, skip the read lock/unlock.
```suggestion
	job.Lock()
	defer job.Unlock()
	job, ok := jobs.log[id]
	if !ok {
		return nil, errors.New("invalid handle")
	}
	if streamType == StreamStdout {
		return job.stdout.Attach(ctx), nil
	}
```

Same thing in all functions below

Reply:
`jobs` RLock/RUnlock is for job cache, while `job` Lock/Unlock is for specific entry in job cache.
```
job.Lock()
defer job.Unlock()
job, ok := jobs.log[id]
```
is unsynchronised access to the map.